### PR TITLE
Renderer support for removing root components or updating parameters on them

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
This is needed as part of #27574 

The core `Renderer` needs to have APIs for being told to:

 * Remove root components, clearing up properly and running all the correct disposal flows
 * Supply new parameters to root components, whether or not the rendering system is currently quiescent, and return a task that completes when we do reach quiescence

It's not necessary to add a new API for "supply new parameters" because the existing `RenderRootComponentAsync` actually does that already. The one thing it didn't support was being called while the system was not yet quiescent, so that's what I've added. It just involves juggling the tasks a little differently.

### CODE REVIEW

File: src/Components/Components/src/RenderTree/Renderer.cs [hunk 1]
```diff
@@ -199,53 +200,67 @@ protected Task RenderRootComponentAsync(int componentId)
...
-                _pendingTasks = null;
+                throw new InvalidOperationException("The specified component is not a root component");
             }
+
+            // This assumes there isn't currently a batch in progress, and will throw if there is.
+            // Currently there's no known scenario where we need to support calling RemoveRootComponentAsync
+            // during a batch, but if a scenario emerges we can add support.
+            _batchBuilder.ComponentDisposalQueue.Enqueue(componentId);
+            _rootComponentsLatestParameters?.Remove(componentId);
+            ProcessRenderQueue();
```
Comment (javiercn) [inline src/Components/Components/src/RenderTree/Renderer.cs:263]: Will this trigger the disposal of all the components in the tree? I remember having to do something more "sophisticated" here (triggering an empty render).

Comment (SteveSandersonMS) [reply src/Components/Components/src/RenderTree/Renderer.cs:263]: It does - disposal is recursive. There's a unit test for that too.

Comment (javiercn) [reply src/Components/Components/src/RenderTree/Renderer.cs:263]: Cool, I'm glad we found a simpler way, the way I did it was a bit convoluted

**Review by javiercn — COMMENTED** [review]
Looks great!

I have to look at things with a fresh brain, but overall looks superb!

File: src/Components/Components/src/RenderTree/Renderer.cs [hunk 2]
```diff
@@ -254,21 +269,45 @@ protected async Task RenderRootComponentAsync(int componentId, ParameterView ini
...
+            // If there's already a loop waiting for quiescence, just join it
+            if (_ongoingQuiescenceTask is not null)
+            {
+                await _ongoingQuiescenceTask;
+                return;
+            }
+
+            try
+            {
+                _ongoingQuiescenceTask ??= ProcessAsynchronousWork();
```
Comment (SteveSandersonMS) [inline src/Components/Components/src/RenderTree/Renderer.cs:285]:
```suggestion
                _ongoingQuiescenceTask = ProcessAsynchronousWork();
```

File: src/Components/Components/src/RenderTree/Renderer.cs [hunk 3]
```diff
@@ -254,21 +269,45 @@ protected async Task RenderRootComponentAsync(int componentId, ParameterView ini
         /// <param name="exception">The <see cref="Exception"/>.</param>
         protected abstract void HandleException(Exception exception);
 
-        private async Task ProcessAsynchronousWork()
+        private Task? _ongoingQuiescenceTask;
```
Comment (SteveSandersonMS) [inline src/Components/Components/src/RenderTree/Renderer.cs:272]: TODO: Move this field to be with other fields

**Comment by SteveSandersonMS** [comment]
@javiercn Is there anything further you want to check here, or is this good to merge? I'm asking because I have a subsequent PR building on this and it would be handy to target `main` with that rather than chaining PRs.

File: src/Components/Components/test/EventCallbackFactoryBinderExtensionsTest.cs [hunk 4]
```diff
@@ -493,7 +493,7 @@ public async Task CreateBinder_NullableDateTimeOffset()
             var expectedValue = new DateTime(2018, 3, 4, 1, 2, 3);
 
             // Act
-            await binder.InvokeAsync(new ChangeEventArgs() { Value = expectedValue.ToString(CultureInfo.InvariantCulture), });
+            await binder.InvokeAsync(new ChangeEventArgs() { Value = expectedValue.ToString(CultureInfo.CurrentCulture), });
```
Comment (javiercn) [inline src/Components/Components/test/EventCallbackFactoryBinderExtensionsTest.cs:496]: This seems to be unrelated, won't `CurrentCulture` cause issues with machines on other locales?

**Review by javiercn — APPROVED** [review]
Looks great!

File: src/Components/Components/test/EventCallbackFactoryBinderExtensionsTest.cs (hunk 4 above)
Comment (SteveSandersonMS) [reply src/Components/Components/test/EventCallbackFactoryBinderExtensionsTest.cs:496]: The test was wrong before. Binding works with culture-specific values. The parsing was already culture-specific, so it was incorrect to test a culture-invariant input, and was failing on my machine (using months for days, and vice-versa).

The only reason it was passing in CI before was due to the cultural imperialism of `CultureInfo.InvariantCulture` being equivalent to US date formatting.

Comment (javiercn) [reply src/Components/Components/test/EventCallbackFactoryBinderExtensionsTest.cs:496]: > the cultural imperialism of `CultureInfo.InvariantCulture` being equivalent to US date formatting

lol, fair enough :).

I only brought it up because we ran into issues in the past in this area.
